### DISCUSSION TITLE
lsim warning for TF systems

### DISCUSSION THREAD

**Discussion by mvaccariunipi** [python-control/python-control#625] [Q&A] [answered]
Dear all,
I have just installed control 0.9.0 and when using lsim (or forced_response) on a transfer function system I got the following warning: "return_x specified for a transfer function system. Internal conversion to state space used; results may meaningless."

Can I skip this warning? Is there another function for simulating transfer function systems that does not require a conversion to state space? 
I cannot figure out if the warning is something I should worry about or not. 

Thanks,

Marco from SIPPY teams

The example tested is the following:

```python
from __future__ import division
from past.utils import old_div
from sippy import functionset as fset
from sippy import *
import numpy as np
import control.matlab as cnt

sampling_time = 1.                                  # [s]
end_time = 400                                      # [s]
npts = int(old_div(end_time, sampling_time)) + 1
Time = np.linspace(0, end_time, npts)

switch_probability = 0.08  # [0..1]
[Usim,_,_] = fset.GBN_seq(npts, switch_probability, Range = [-1, 1])

white_noise_variance = [0.01]
e_t = fset.white_noise_var(Usim.size, white_noise_variance)[0]


DEN = [1., -2.21, 1.7494, -0.584256, 0.0684029, 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]
NUM = [1.5, -2.07, 1.3146]

g_sample = cnt.tf(NUM, DEN, sampling_time)

Y1, Time, Xsim = cnt.lsim(g_sample, Usim, Time)
```

**Comment by sawyerbfuller** [comment]
You can ignore that warning if you don't intend to make use of x. x is the "internal state" of the system, and depends on the specific realization of your state-space representation of the transfer function you provided.

**Comment by mvaccariunipi** [comment]
Hi @ sawyerbfuller and thanks for the prompt answer. Is there a way to not show the warning? Otherwise I'll just ignore it. 

Thanks again,

Marco

**Comment by sawyerbfuller** [comment, marked as answer]
Hi Marco, 

one way is to use the more pythonic forced_response, which gives you more control, and in particular, I believe you can ask it to not give you x.

**Reply by murrayrm** [reply]
This is the best approach.  The `forced_response` function does not return the state unless you set `return_x` explicitly (the default is to just return the output).

The reason that you get this error with `lsim` is that the MATLAB compatibility package matches MATLABs call signature and so returns the state by default.  However, for a transfer function there is not a well defined state (it depends on the realization) and so this issues a warning.  

Here's the code for `control.matlab.lsim`:
```
from ..timeresp import forced_response

# Switch output argument order and transpose outputs (and always return x)
out = forced_response(sys, T, U, X0, return_x=True, transpose=True)
return out[1], out[0], out[2]
```
If you just replicate this call (as @sawyerbfuller suggested), you can set `return_x=False` (which is the default).

You can also turn off the warning using a warnings filter.

**Reply by mvaccariunipi** [reply]
Ok, thanks @murrayrm and @sawyerbfuller  for the suggestions. I will think about it.

Bests wishes to all.

**Comment by CPCLAB-UNIPI** [comment]
Dear @murrayrm and @sawyerbfuller,
following your suggestions we tried to replicate the direct call to forced_response but failed because of the following error:

ValueError: Parameter T: time values must be equally spaced.

Digging into the code we found the error line and tested the commands on the input we passed:
T = np.linspace(0, 400, 401)

dt = T[1] - T[0] = 1 line 340 of timeresp.py
np.allclose(T[1:] - T[:-1], dt) = True

Is it possible that the function _check_convert_array "does something wrong" to the array produced by linspace so
that the check at line 341 returns False?

Can you help us with this issue?

Thanks,

SIPPY team

**Reply by murrayrm** [reply]
Are you calling `forced_response` with the right arguments?  Note that `lsim` has signature `lsim(sys, U, T)` but `forced_response` has signature `forced_response(sys, T, U)`.

The following worked for me:
```
import numpy as np
import control as ct
Time = np.linspace(0, 400, 401)

DEN = [1., -2.21, 1.7494, -0.584256, 0.0684029, 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]
NUM = [1.5, -2.07, 1.3146]
g_sample = ct.tf(NUM, DEN, 1)

Usim = np.sin(Time)
Y1, Time, Xsim = ct.forced_response(g_sample, Time, Usim, return_x=True)
```

**Reply by bnavigator** [reply]
Hi,

I am not one of the mentioned contributors, but I hope I can help.

`_check_convert_array`  does not change values or the length of the array, only the shape. So it is unlikely the cause of the error.

Note that we already changed the code to check for a correct time vector between version 0.9.0 and the current development commit. I tried to reproduce your provided code with python-control version 0.9.0 and the current development commit. Both return `True`:

```python
import control
import numpy as np

T = np.linspace(0, 400, 401)
_T = control.timeresp._check_convert_array(T, [('any',), (1, 'any')],
    'Parameter ``T``: ', squeeze=True, transpose=False)
print(np.allclose(_T[1:] - _T[:-1], _T[1] - _T[0]))
```

**Reply by CPCLAB-UNIPI** [reply]
Dear @murrayrm and @bnavigator,
thanks for the clarifications. 
We managed to work out the call to `forced_response` function. 
Thanks for the help.

Best regards,

SIPPY team
